File: Arrays and Vectors/MinimumSwaps.cpp
#include<bits/stdc++.h>
using namespace std;

bool compare(pair<int, int> p1, pair<int, int> p2) {
    return p1.first < p2.first;
}

int min_swaps(vector<int> & v) {
    int n = v.size();
    
    vector<pair<int, int>> nv;
    
    for(int i=0; i<n; i++) {
        nv.push_back({v[i], i});
    }
    
    sort(nv.begin(), nv.end(), compare);
    
    vector<bool> visited(n, false);
    int total = 0;
    
    for(int i=0; i<n; i++) {
        if(visited[i] || v[i] == nv[i].first) {
            continue;
        }
        
        int cycles = 0;
        int node = i;
        while(!visited[node]) {
            visited[node] = true;
            node = nv[node].second;
            cycles ++;
        }
        
        total += cycles-1;
    }
    
    return total;
}

int main() {
    vector<int> v = {4,5,3,2,1};
    int n = v.size();
    cout<<min_swaps(v)<<endl;
}

File: Arrays and Vectors/MinimumSwaps.js
function MinimumSwaps(nums) {
  let n = nums.length;

  let pNums = [];
  for (let i = 0; i < n; i++) {
    pNums.push([nums[i], i]);
  }

  pNums.sort((p1, p2) => p1[0] - p2[0]);

  let visited = nums.map((_) => false);

  let swaps = 0;

  for (let i = 0; i < n; i++) {
    if (pNums[i][0] === nums[i] || visited[i]) {
      continue;
    }

    let cycles = 0;
    let node = i;
    while (!visited[node]) {
      visited[node] = true;
      node = pNums[node][1];
      cycles++;
    }

    swaps += cycles - 1;
  }

  return swaps;
}

const nums = [5, 3, 4, 2, 1];
console.log(MinimumSwaps(nums));
